### PR TITLE
dynamic scoring for LM

### PR DESCRIPTION
**Pull request by l-k-11235**
This PRs adapts the ScoringPreparator for LM architecture and fixes the ignore_prompt method of the loss for LM validation with left padding

### CODE REVIEW

File: eole/predict/inference.py [hunk 1]
```diff
@@ -436,7 +436,7 @@ def _process_bucket(bucket_predictions):
             batch_data = self.predict_batch(batch, attn_debug)
 
             predictions = prediction_builder.from_batch(batch_data)
-            is_seq2seq = hasattr(self.model, "encoder") and hasattr(self.model, "decoder")
+            is_seq2seq = self.model.encoder is not None and self.model.decoder is not None
```
Comment (vince62s) [inline eole/predict/inference.py:439]: what is the edge where this is different from previous code?

File: eole/transforms/tokenize_id.py [hunk 2]
```diff
@@ -92,6 +92,9 @@ def tokenize_string(self, string, side="src", is_train=False):
             kwargs = {"max_length": self.max_length, "truncation": True}
         else:
             kwargs = {}
+        string = string.replace(DefaultTokens.SEP, "\n").replace(
+            DefaultTokens.MASK_BEFORE, self.tokenizers[side].pad_token
+        )
```
Comment (vince62s) [inline eole/transforms/tokenize_id.py:97]: shouldn't we handle it the same way for other tokenizers ?

File: eole/utils/loss.py [hunk 3]
```diff
@@ -255,9 +255,11 @@ def ignore_prompt(self, batch):
...
         """
         # Create a mask with zeros at prompt positions and ones at answer postions.
-        mask = batch["src"].squeeze(dim=-1) == self.padding_idx
-        mask = torch.cumsum(mask.int(), 1)
-        # Apply the mask on the target side.
+        mask = batch["tgt"].squeeze(dim=-1) == self.padding_idx
+        mask = mask.cumsum(dim=1)
+        row_max = mask.max(dim=1, keepdim=True).values
+        mask = torch.where(mask < row_max, 0, mask)
+        mask = torch.where(mask >= row_max, 1, mask)
```
Comment (vince62s) [inline eole/utils/loss.py:262]: I think you could use simpler stuff like:
mask = (batch["tgt"].squeeze(dim=-1) == self.padding_idx).cumsum(dim=1)
mask = (mask >= mask.max(dim=1, keepdim=True).values)
BUT CHECK FIRST

File: eole/utils/scoring_utils.py [hunk 4]
```diff
@@ -54,9 +54,17 @@ def translate(self, model, gpu_rank, step):
         # (take 'inference' field of config if exists?)
         # Set "default" translation options on empty cfgfile
         self.config.training.num_workers = 0
+        is_seq2seq = model.encoder is not None and model.decoder is not None
+        print("is_seq2seq", is_seq2seq)
```
Comment (vince62s) [inline eole/utils/scoring_utils.py:58]: remove print

File: eole/utils/scoring_utils.py [hunk 5]
```diff
@@ -54,9 +54,17 @@ def translate(self, model, gpu_rank, step):
...
+        if not is_seq2seq:
+            if "insert_mask_before_placeholder" in self.config.transforms:
+                self.response_patterns = self.config.transforms_configs.insert_mask_before_placeholder.response_patterns
+        else:
+            self.response_patterns = None
+
         predict_config = PredictConfig(
             model_path=["dummy"],
-            src=self.config.data["valid"].path_src,
+            src="dummy",
```
Comment (vince62s) [inline eole/utils/scoring_utils.py:67]: not sure to understand, don't we lose the the valid src ?

File: eole/utils/scoring_utils.py [hunk 6]
```diff
@@ -67,33 +75,60 @@ def translate(self, model, gpu_rank, step):
...
-            model_config,
-            device_id=gpu_rank,
-            global_scorer=scorer,
-            report_align=predict_config.report_align,
-            report_score=False,
-            logger=None,
-        )
+
+        if is_seq2seq:
+            translator = Translator.from_config(  # we need to review opt/config stuff in translator
```
Comment (vince62s) [inline eole/utils/scoring_utils.py:80]: maybe lets rename "translator" by "predictor" and below too

File: eole/utils/scoring_utils.py [hunk 7]
```diff
@@ -102,6 +137,11 @@ def translate(self, model, gpu_rank, step):
         # ########### #
         # Predictions #
         # ########### #
+        if not is_seq2seq:
+            translator.id_tokenization = True
```
Comment (vince62s) [inline eole/utils/scoring_utils.py:141]: well not sure to understand, what about when we don't use HF tokenizers ?

File: eole/predict/inference.py (hunk 1 above)
Comment (l-k-11235) [reply eole/predict/inference.py:439]: In my test with Eurollm self.encoder was equal to None.

File: eole/transforms/tokenize_id.py (hunk 2 above)
Comment (l-k-11235) [reply eole/transforms/tokenize_id.py:97]: I have only tested for that one (with eurollm) at this point; an error will be raised with the others.

File: eole/utils/scoring_utils.py (hunk 5 above)
Comment (l-k-11235) [reply eole/utils/scoring_utils.py:67]: No, it is passed to the inference iterator below.

File: eole/utils/scoring_utils.py (hunk 7 above)
Comment (l-k-11235) [reply eole/utils/scoring_utils.py:141]: It is supposed to work as well with others tokenizers, no ?

File: eole/utils/loss.py (hunk 3 above)
Comment (l-k-11235) [reply eole/utils/loss.py:262]: It works fine !

File: eole/transforms/tokenize_id.py (hunk 2 above)
Comment (l-k-11235) [reply eole/transforms/tokenize_id.py:97]: In fact the logic to apply id_tokenization already exists here
https://github.com/eole-nlp/eole/blob/40ed6ea8c967107ac7da33aa926433845ee2d268/eole/predict/inference.py#L196-L200
The point is that the "filtertoolong" transform shouldn't be applied after "huggingface_tokenize". See https://github.com/eole-nlp/eole/issues/191
I remove this.

File: eole/utils/scoring_utils.py [hunk 8]
```diff
@@ -42,58 +42,94 @@ def translate(self, model, gpu_rank, step):
...
+                model,
+                self.vocabs,
+                predict_config,
+                model_config,
+                device_id=gpu_rank,
+                global_scorer=scorer,
+                report_align=predict_config.report_align,
+                report_score=False,
+                logger=None,
+            )
```
Comment (francoishernandez) [inline eole/utils/scoring_utils.py:103]: Maybe cleaner to just define a `predictor_class` in the condition, and call `predictor_class(*)` once, since they should have the same signature.

**Review by francoishernandez — COMMENTED** [review]
It seems a bit weird to have such very specific changes in `scoring_utils` (`response_patterns` handling mostly), but I'm not sure how this can be properly factorized with the current structure. Let's keep it for now and we might reconsider later.
